src/simulators/statevector/statevector_state.hpp

```diff
@@ -634,6 +655,82 @@ void State<statevec_t>::snapshot_matrix_expval(const Operations::Op &op,
   BaseState::qreg_.revert(false);
 }
 
+template <class statevec_t>
+void State<statevec_t>::snapshot_density_matrix(const Operations::Op &op,
+                                                ExperimentData &data,
+                                                SnapshotDataType type) {
+  cmatrix_t reduced_state;
+
+  // Check if tracing over all qubits
+  if (op.qubits.empty()) {
+    reduced_state = cmatrix_t(1, 1);
+    reduced_state[0] = BaseState::qreg_.norm();
+  } else {
+    reduced_state = density_matrix(op.qubits);
+  }
+
+  // Add density matrix to result data
+  switch (type) {
+    case SnapshotDataType::average:
+      data.add_average_snapshot("density_matrix", op.string_params[0],
+                            BaseState::creg_.memory_hex(), std::move(reduced_state), false);
+      break;
+    case SnapshotDataType::average_var:
+      data.add_average_snapshot("density_matrix", op.string_params[0],
+                            BaseState::creg_.memory_hex(), std::move(reduced_state), true);
+      break;
+    case SnapshotDataType::pershot:
+      data.add_pershot_snapshot("density_matrix", op.string_params[0], std::move(reduced_state));
+      break;
+  }
+}
+
+template <class statevec_t>
+cmatrix_t State<statevec_t>::density_matrix(const reg_t &qubits) {
+  return vec2density(qubits, BaseState::qreg_.data());
+}
+
+#ifdef AER_THRUST_SUPPORTED
+template <>
+cmatrix_t State<QV::QubitVectorThrust<float>>::density_matrix(const reg_t &qubits) {
+  return vec2density(qubits, BaseState::qreg_.vector());
+}
+
+template <>
+cmatrix_t State<QV::QubitVectorThrust<double>>::density_matrix(const reg_t &qubits) {
+  return vec2density(qubits, BaseState::qreg_.vector());
+}
+#endif
+
+template <class statevec_t>
+template <class T>
+cmatrix_t State<statevec_t>::vec2density(const reg_t &qubits, const T &vec) {
+  const size_t N = qubits.size();
+  const size_t DIM = 1ULL << N;
+  auto qubits_sorted = qubits;
+  std::sort(qubits_sorted.begin(), qubits_sorted.end());
+
+  // Return full density matrix
+  cmatrix_t densmat(DIM, DIM);
+  if ((N == BaseState::qreg_.num_qubits()) && (qubits == qubits_sorted)) {
+    
+    for (size_t row = 0; row < DIM; ++row)
```

Comment:
I think we could use OpenMP pragmas here to parallelize these loops.

Reply:
This is a reduction so for one you would have to define a custom reduction type for OMP here, and I think that would result in memory for the matrix being duplicated across threads. For these types of snapshots which can be very large in memory usage I think minimizing memory is more important than speed.